File: general-interview-prep/golang/isCompleteTree.go
package main

import "fmt"
type TreeNode struct {
     Val int
     Left *TreeNode
     Right *TreeNode
 }

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func isCompleteTree(root *TreeNode) bool {
    if root == nil {
        return true
    }

    gap := false
    queue := []*TreeNode{root}

    for len(queue) > 0 {
        next := []*TreeNode{}

        for _, current := range queue {
            if current.Left != nil && gap == true {
                return false
            }

            if current.Left != nil {
                next = append(next, current.Left)
            } else {
                gap = true
            }

            if current.Right != nil && gap == true {
                return false
            }

            if current.Right != nil {
                next = append(next, current.Right)
            } else {
                gap = true
            }
        }

        queue = next
    }

    return true
}


func main() {
  fmt.Println(isCompleteTree(&TreeNode{Val: 1, Left: &TreeNode{Val: 2, Left: &TreeNode{Val: 4}, Right: &TreeNode{Val: 5}}, Right: &TreeNode{Val: 3, Left: &TreeNode{Val: 6}}}))
}

File: general-interview-prep/javascript/isCompleteTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */


 var isCompleteTree = function(root) {
  let arr = [];
  if (!root) {
      return false;
  }
  
  let gap = false;
  let queue = [root];
  
  while (queue.length) {
      let next = [];
      for (let i = 0; i < queue.length; i++) {
          let current = queue[i];
          if (current.left !== null && gap ) {
              return false;
          }
          
          if (current.left) {
              next.push(current.left);
          } else {
              gap = true;
          }
          
          if (current.right !== null && gap) {
              return false;
          }
          
          if (current.right) {
              next.push(current.right);
          } else {
              gap = true;
          }
      }
      
      queue = next;
  }
  
  return true
};